core/src/engine/impl/mutex_impl.hpp

```diff
@@ -0,0 +1,166 @@
+
+#include <atomic>
+
+#include <userver/engine/deadline.hpp>
+
+#include <userver/utils/assert.hpp>
+
+#include <engine/impl/wait_list.hpp>
+#include <engine/impl/wait_list_light.hpp>
+#include <engine/task/task_context.hpp>
+
+USERVER_NAMESPACE_BEGIN
+
+namespace engine::impl {
+
+template <class Waiters>
+class MutexImpl {
+ public:
+  MutexImpl();
+  ~MutexImpl();
+
+  MutexImpl(const MutexImpl&) = delete;
+  MutexImpl(MutexImpl&&) = delete;
+  MutexImpl& operator=(const MutexImpl&) = delete;
+  MutexImpl& operator=(MutexImpl&&) = delete;
+
+  void lock();
+
+  void unlock();
+
+  bool try_lock();
+
+  bool try_lock_until(Deadline deadline);
+
+ private:
+  class MutexWaitStrategy;
+
+  bool LockFastPath(TaskContext&);
+  bool LockSlowPath(TaskContext&, Deadline);
+
+  std::atomic<TaskContext*> owner_;
+  Waiters lock_waiters_;
+};
+
+
+
+template<>
+class MutexImpl<WaitList>::MutexWaitStrategy final : public WaitStrategy {
+ public:
+  MutexWaitStrategy(MutexImpl<WaitList>& mutex, TaskContext& current,
+                    Deadline deadline)
+      : WaitStrategy(deadline),
+        mutex_(mutex),
+        current_(current),
+        waiter_token_(mutex_.lock_waiters_),
+        lock_(mutex_.lock_waiters_) {}
+
+  void SetupWakeups() override {
+    mutex_.lock_waiters_.Append(lock_, &current_);
+    lock_.unlock();
+  }
+
+  void DisableWakeups() override {
+    lock_.lock();
+    mutex_.lock_waiters_.Remove(lock_, current_);
+  }
+
+ private:
+  MutexImpl<WaitList>& mutex_;
+  TaskContext& current_;
+  const WaitList::WaitersScopeCounter waiter_token_;
+  WaitList::Lock lock_;
+};
+
+template<>
+class MutexImpl<WaitListLight>::MutexWaitStrategy final : public WaitStrategy {
+ public:
+  MutexWaitStrategy(MutexImpl<WaitListLight>& mutex, TaskContext& current,
+                    Deadline deadline)
+      : WaitStrategy(deadline),
+        mutex_(mutex),
+        current_(current) {}
+
+  void SetupWakeups() override {
+    mutex_.lock_waiters_.Append(&current_);
+    if (!mutex_.owner_.load()) mutex_.lock_waiters_.WakeupOne();
+  }
+
+  void DisableWakeups() override {
+    mutex_.lock_waiters_.Remove(current_);
+  }
+
+ private:
+  MutexImpl<WaitListLight>& mutex_;
+  TaskContext& current_;
+};
+
+template <class Waiters>
+MutexImpl<Waiters>::MutexImpl() : owner_(nullptr) {}
+
+template <class Waiters>
+MutexImpl<Waiters>::~MutexImpl() { UASSERT(!owner_); }
+
+template <class Waiters>
+bool MutexImpl<Waiters>::LockFastPath(TaskContext& current) {
+  TaskContext* expected = nullptr;
+  return owner_.compare_exchange_strong(expected, &current,
+                                        std::memory_order_acquire);
+}
+
+template <class Waiters>
+bool MutexImpl<Waiters>::LockSlowPath(TaskContext& current, Deadline deadline) {
+  TaskContext* expected = nullptr;
+
+  const engine::TaskCancellationBlocker block_cancels;
+  MutexWaitStrategy wait_manager(*this, current, deadline);
+  while (!owner_.compare_exchange_strong(expected, &current,
```

Comment:
issue: This loop is needed for `Mutex`, but not for `SingleWaitingTaskMutex`:
- in `Mutex`, there may be multiple waiters, and under certain circumstances the lock may be stolen from the woken up task
- in `SingleWaitingTaskMutex`, when a waiter is woken up with `WakeupSource::kWaitList`, it will acquire the lock for sure

We can even use a relaxed store, because the other threads/tasks still can't begin waiting while we haven't exited `LockSlowPath`. Insert an `UASSERT` for vacancy to be sure.

This means `SingleWaitingTaskMutex` can't share the implementation with `Mutex`.

Also, CAS is not needed in `LockFastPath` iff it is performed inside `lock` (not `try_lock`). It is enough to relaxed-`load` and release-`store`, if `nullptr`.

These changes should shave off something like 20ns.

Reply:
Benchmark before suggested optimizations 
```
---------------------------------------------------------------------------------------------
Benchmark                                                      Time           CPU Iterations
---------------------------------------------------------------------------------------------
mutex_coro_lock                                               13 ns         13 ns   52964899
mutex_std_lock                                                14 ns         14 ns   52556104
single_waiting_task_mutex_lock                                14 ns         14 ns   49777070
mutex_coro_unlock                                             11 ns         11 ns   58282822
mutex_std_unlock                                              13 ns         13 ns   51978139
single_waiting_task_mutex_unlock                              22 ns         22 ns   25760581
mutex_coro_contention/1                                       14 ns         14 ns   51055388 locks=72.2736M/s locks-per-thread=72.2736M/s
mutex_coro_contention/2                                      111 ns         94 ns    6951686 locks=24.6446M/s locks-per-thread=12.3223M/s
mutex_std_contention/1                                        19 ns         19 ns   32611076 locks=53.3898M/s locks-per-thread=53.3898M/s
mutex_std_contention/2                                        89 ns         72 ns   10000000 locks=28.5492M/s locks-per-thread=14.2746M/s
single_waiting_task_mutex_contention/1                        27 ns         27 ns   26160498 locks=37.7132M/s locks-per-thread=37.7132M/s
single_waiting_task_mutex_contention/2                       129 ns        129 ns    4861651 locks=15.5616M/s locks-per-thread=7.78081M/s
mutex_coro_contention_with_payload/1                         110 ns        110 ns    5751754 locks=9.07964M/s locks-per-thread=9.07964M/s
mutex_coro_contention_with_payload/2                         571 ns        520 ns    1000000 locks=3.4707M/s locks-per-thread=1.73535M/s
mutex_std_contention_with_payload/1                          121 ns        121 ns    5902142 locks=8.27959M/s locks-per-thread=8.27959M/s
mutex_std_contention_with_payload/2                          555 ns        451 ns    1527837 locks=4.22651M/s locks-per-thread=2.11325M/s
single_waiting_task_mutex_contention_with_payload/1          125 ns        125 ns    5425557 locks=8.02919M/s locks-per-thread=8.02919M/s
single_waiting_task_mutex_contention_with_payload/2          415 ns        413 ns    1855987 locks=4.74809M/s locks-per-thread=2.37405M/s

```

Reply:
Sorry, my optimizations are invalid, because the thread that has just released the lock is allowed to re-acquire it immediately.